### PR TITLE
Refactor iterator to separate into HTTP/GAX iterators

### PR DESCRIPTION
**Pull request by dhermes**
**NOTE**: I'm happy to split this PR into tiny PRs if it makes review easy, though the commits are all pretty compact, so it should be reviewable commit-by-commit.

### CODE REVIEW

**Comment by jonparrott** [comment]
This conceptually LGTM, I'll defer @tseaver for implementation review.

**Comment by dhermes** [comment]
@tseaver Can I get an amen?

File: core/google/cloud/iterator.py [hunk 1]
```diff
@@ -166,7 +166,7 @@ def next(self):
     __next__ = next
 
 
-class Iterator(object):
+class HTTPIterator(object):
     """A generic class for iterating through Cloud JSON APIs list responses.
```
Comment (tseaver) [inline core/google/cloud/iterator.py]: Good naming, given that it wraps around a path.

File: core/unit_tests/test_iterator.py [hunk 2]
```diff
@@ -112,6 +112,7 @@ def test_constructor(self):
         self.assertIs(iterator.client, client)
         self.assertEqual(iterator.max_results, max_results)
         self.assertEqual(list(iterator._page_iter), [])
+        self.assertFalse(iterator._page_increment)
```
Comment (tseaver) [inline core/unit_tests/test_iterator.py]: Shouldn't there be a matching assertion for when it becomes True?

File: pubsub/google/cloud/pubsub/_gax.py [hunk 3]
```diff
@@ -569,24 +564,3 @@ def _item_to_topic(iterator, resource):
...
-    :param page_iter: The iterator to wrap.
-
-    :type iterator: :class:`~google.cloud.iterator.Iterator`
-    :param iterator: The iterator that owns each page.
-    """
-    for items in page_iter:
-        page = Page(iterator, items, _item_to_topic)
-        iterator.next_page_token = page_iter.page_token or None
-        iterator.num_results += page.num_items
-        yield page
```
Comment (tseaver) [inline pubsub/google/cloud/pubsub/_gax.py]: "And there was much rejoicing."

File: core/google/cloud/iterator.py [hunk 4]
```diff
@@ -393,6 +407,24 @@ def __init__(self, client, page_iter, item_to_value, max_results=None):
...
+        """
+        try:
+            items = six.next(page_iter)
+            page = Page(self, items, self._item_to_value)
+            self.next_page_token = page_iter.page_token or None
+            return page
+        except StopIteration:
+            return None
+
     def _wrap_gax(self, page_iter):
```
Comment (tseaver) [inline core/google/cloud/iterator.py]: Would it be better to rename `_wrap_gax` -> `_gax_page_iter`, to indicate its parallel w/ `_http_page_iter`?

File: core/unit_tests/test_iterator.py [hunk 5]
```diff
@@ -400,8 +400,6 @@ def test__get_next_page_response_new_no_token_in_response(self):
         iterator = self._makeOne(client, path, None)
         response = iterator._get_next_page_response()
         self.assertEqual(response['items'], [{'name': key1}, {'name': key2}])
-        self.assertEqual(iterator.page_number, 1)
-        self.assertEqual(iterator.next_page_token, token)
```
Comment (tseaver) [inline core/unit_tests/test_iterator.py]: Do we need assertions that thes values get updated in `_next_page` / `_http_page_iter` / `_wrap_gax`?

Comment (dhermes) [inline core/unit_tests/test_iterator.py]: @tseaver Is this question still relevant? Especially now that `_http_page_iter` and `_wrap_gax` don't exist?

**Comment by dhermes** [comment]
@tseaver Sorry for the bait and switch at the end. It didn't become clear that a `_next_page` could do all the work until all pieces were moved. I considered rewriting the history but hoped it'd be clear enough, my bad.

File: core/unit_tests/test_iterator.py (hunk 5 above)
Comment (tseaver) [inline core/unit_tests/test_iterator.py]: I'll defer to your judgement: as I said, the commit-by-commit review falls over when the endgame completely scraps the earlier work.

**Pull request merged**